src/ensemble_scheduler/ensemble_scheduler.cc

```diff
@@ -1033,6 +1040,57 @@ EnsembleContext::ReshapeTensorDims(
   return res;
 }
 
+// Caching function
+void
+EnsembleContext::CacheEnsembleTopLevelRequest(
+    std::unique_ptr<InferenceResponse>& response)
+{
+  const std::string key = request_tracker_->Request()->CacheKey();
+  const bool is_key_set = request_tracker_->Request()->CacheKeyIsSet();
+
+#ifdef TRITON_ENABLE_STATS
+  info_->ensemble_end_ns = CaptureTimeNs();
+  const uint64_t lookup_end_ns =
+      request_tracker_->Request()->CacheLookupEndNs();
+  const uint64_t lookup_start_ns =
+      request_tracker_->Request()->CacheLookupStartNs();
+#endif
+
+  if (!is_key_set) {
+    LOG_ERROR << "Request cache key was not set correctly.";
+  }
+
+  auto cache = is_->CacheManager()->Cache();
+#ifdef TRITON_ENABLE_STATS
+  const uint64_t insert_start_ns = CaptureTimeNs();
+#endif
+  auto status = cache->Insert(response.get(), key);
+  // LOG_VERBOSE(1) << response->ModelName() << " " <<
+  // response->ActualModelVersion() << " " << response.;
+  if (!status.IsOk()) {
+    LOG_ERROR << "Failed to insert key [" << key
```

Comment:
Do we have any tests cases that expect a failed insertion to make sure server handles it gracefully and falls back to normal inference?

Reply:
We don't have any tests. I can add a testcase for this case.

Reply:
Test added in L0_response_cache